README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# About `scorecard`

[![Travis-CI Build
Status](https://travis-ci.org/jjchern/scorecard.svg?branch=master)](https://travis-ci.org/jjchern/scorecard)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/jjchern/scorecard?branch=master&svg=true)](https://ci.appveyor.com/project/jjchern/scorecard)

The `scorecard` package includes processed datasets from the [College
Scorecard](https://collegescorecard.ed.gov), 1996-2017.

The Scorecard datasets are imbalance panels at the
colleges-by-school-year level. The data was last updated in 2019. See
the [changelog](https://collegescorecard.ed.gov/data/changelog/) for
more details.

The following datasets are available:

  - `scorecard::mf1996_97`
  - `scorecard::mf1997_98`
  - `scorecard::mf1998_99`
  - `scorecard::mf1999_00`
  - `scorecard::mf2000_01`
  - `scorecard::mf2001_02`
  - `scorecard::mf2002_03`
  - `scorecard::mf2003_04`
  - `scorecard::mf2004_05`
  - `scorecard::mf2005_06`
  - `scorecard::mf2006_07`
  - `scorecard::mf2007_08`
  - `scorecard::mf2008_09`
  - `scorecard::mf2009_10`
  - `scorecard::mf2010_11`
  - `scorecard::mf2011_12`
  - `scorecard::mf2012_13`
  - `scorecard::mf2013_14`
  - `scorecard::mf2014_15`
  - `scorecard::mf2015_16`
  - `scorecard::mf2016_17`
  - `scorecard::mf2017_18`
  - `scorecard::codebook`
  - `scorecard::cohort_map`

The following functions are implemented:

  - `scorecard::attach_var_label()`: Assign variable labels.
  - `scorecard::attach_val_label()`: Assign value labels for certain
    variables

# Related R Package

[Benjamin Skinner](https://github.com/btskinner) has created a wonderful
R client [`rscorecard`](http://btskinner.me/rscorecard/) for the
[College Scorecard GET
API](https://collegescorecard.ed.gov/data/documentation/). If you’re
interested in getting some specific variables quickly, I suggest using
the `rscorecard` package.

# Installation

You can also download the datasets as an R package. It might take a
while to install and load into memory. To download the most recent
stable release, use

``` r
# install.packages("devtools")
devtools::install_github("jjchern/scorecard@v0.4.0")

# To uninstall the package, use:
# remove.packages("scorecard")
```

# Examples

## Loading the merged file for school year 2014-15

All datasets are tibbles:

``` r
library(tidyverse)
scorecard::mf2014_15
#> # A tibble: 7,703 x 1,660
#>    mf_year unitid opeid opeid6 instnm city  stabbr zip   sch_deg    main
#>    <chr>   <chr>  <chr> <chr>  <chr>  <chr> <chr>  <chr>   <dbl> <dbl+l>
#>  1 2014-15 100654 0010… 001002 Alaba… Norm… AL     35762       3 1 [Mai…
#>  2 2014-15 100663 0010… 001052 Unive… Birm… AL     3529…       3 1 [Mai…
#>  3 2014-15 100690 0250… 025034 Amrid… Mont… AL     3611…       3 1 [Mai…
#>  4 2014-15 100706 0010… 001055 Unive… Hunt… AL     35899       3 1 [Mai…
#>  5 2014-15 100724 0010… 001005 Alaba… Mont… AL     3610…       3 1 [Mai…
#>  6 2014-15 100751 0010… 001051 The U… Tusc… AL     3548…       3 1 [Mai…
#>  7 2014-15 100760 0010… 001007 Centr… Alex… AL     35010       2 1 [Mai…
#>  8 2014-15 100812 0010… 001008 Athen… Athe… AL     35611       3 1 [Mai…
#>  9 2014-15 100830 0083… 008310 Aubur… Mont… AL     3611…       3 1 [Mai…
#> 10 2014-15 100858 0010… 001009 Aubur… Aubu… AL     36849       3 1 [Mai…
#> # … with 7,693 more rows, and 1,650 more variables: numbranch <dbl>,
#> #   preddeg <dbl+lbl>, highdeg <dbl+lbl>, control <dbl+lbl>,
#> #   st_fips <dbl+lbl>, region <dbl+lbl>, adm_rate <dbl>,
#> #   adm_rate_all <dbl>, satvr25 <dbl>, satvr75 <dbl>, satmt25 <dbl>,
#> #   satmt75 <dbl>, satwr25 <dbl>, satwr75 <dbl>, satvrmid <dbl>,
#> #   satmtmid <dbl>, satwrmid <dbl>, actcm25 <dbl>, actcm75 <dbl>,
#> #   acten25 <dbl>, acten75 <dbl>, actmt25 <dbl>, actmt75 <dbl>,
#> #   actwr25 <dbl>, actwr75 <dbl>, actcmmid <dbl>, actenmid <dbl>,
#> #   actmtmid <dbl>, actwrmid <dbl>, sat_avg <dbl>, sat_avg_all <dbl>,
#> #   pcip01 <dbl>, pcip03 <dbl>, pcip04 <dbl>, pcip05 <dbl>, pcip09 <dbl>,
#> #   pcip10 <dbl>, pcip11 <dbl>, pcip12 <dbl>, pcip13 <dbl>, pcip14 <dbl>,
#> #   pcip15 <dbl>, pcip16 <dbl>, pcip19 <dbl>, pcip22 <dbl>, pcip23 <dbl>,
#> #   pcip24 <dbl>, pcip25 <dbl>, pcip26 <dbl>, pcip27 <dbl>, pcip29 <dbl>,
#> #   pcip30 <dbl>, pcip31 <dbl>, pcip38 <dbl>, pcip39 <dbl>, pcip40 <dbl>,
#> #   pcip41 <dbl>, pcip42 <dbl>, pcip43 <dbl>, pcip44 <dbl>, pcip45 <dbl>,
#> #   pcip46 <dbl>, pcip47 <dbl>, pcip48 <dbl>, pcip49 <dbl>, pcip50 <dbl>,
#> #   pcip51 <dbl>, pcip52 <dbl>, pcip54 <dbl>, cip01cert1 <dbl+lbl>,
#> #   cip01cert2 <dbl+lbl>, cip01assoc <dbl+lbl>, cip01cert4 <dbl>,
#> #   cip01bachl <dbl>, cip03cert1 <dbl>, cip03cert2 <dbl>,
#> #   cip03assoc <dbl>, cip03cert4 <dbl>, cip03bachl <dbl>,
#> #   cip04cert1 <dbl>, cip04cert2 <dbl>, cip04assoc <dbl>,
#> #   cip04cert4 <dbl>, cip04bachl <dbl>, cip05cert1 <dbl>,
#> #   cip05cert2 <dbl>, cip05assoc <dbl>, cip05cert4 <dbl>,
#> #   cip05bachl <dbl>, cip09cert1 <dbl>, cip09cert2 <dbl>,
#> #   cip09assoc <dbl>, cip09cert4 <dbl>, cip09bachl <dbl>,
#> #   cip10cert1 <dbl>, cip10cert2 <dbl>, cip10assoc <dbl>,
#> #   cip10cert4 <dbl>, cip10bachl <dbl>, cip11cert1 <dbl>, …
```

## Working with variable and value labels

All the datasets have variable labels attached, which can be viewed in
RStudio’s Data Viewer:

``` r
View(scorecard::mf2014_15)
```

![](README-files/RStudio-Data-Viewer.png)

You can also use the `labelled` package:

``` r
scorecard::mf2014_15 %>% 
  select(1:8) %>% 
  labelled::var_label()
#> $mf_year
#> NULL
#> 
#> $unitid
#> [1] "Unit ID for institution"
#> 
#> $opeid
#> [1] "8-digit OPE ID for institution"
#> 
#> $opeid6
#> [1] "6-digit OPE ID for institution"
#> 
#> $instnm
#> [1] "Institution name"
#> 
#> $city
#> [1] "City"
#> 
#> $stabbr
#> [1] "State postcode"
#> 
#> $zip
#> [1] "ZIP code"
```

Or work with the codebook directly:

``` r
## Show variable labels
scorecard::codebook %>% 
  select(var_name, var_label)
#> # A tibble: 2,231 x 2
#>    var_name     var_label                                 
#>    <chr>        <chr>                                     
#>  1 unitid       Unit ID for institution                   
#>  2 opeid        8-digit OPE ID for institution            
#>  3 opeid6       6-digit OPE ID for institution            
#>  4 instnm       Institution name                          
#>  5 city         City                                      
#>  6 stabbr       State postcode                            
#>  7 zip          ZIP code                                  
#>  8 accredagency Accreditor for institution                
#>  9 insturl      URL for institution's homepage            
#> 10 npcurl       URL for institution's net price calculator
#> # … with 2,221 more rows

## Build a small function to shown value labels
show_val_label = . %>% {
    filter(scorecard::codebook, var_name == .) %>% 
    mutate(val_label = glue::glue("{val_label}  = {value}")) %>% 
    distinct(val_label)}

## Show value labels:
show_val_label("curroper")
#> # A tibble: 2 x 1
#>   val_label                                               
#>   <glue>                                                  
#> 1 Not currently certified as an operating institution  = 0
#> 2 Currently certified as operating  = 1
show_val_label("distanceonly")
#> # A tibble: 2 x 1
#>   val_label                       
#>   <glue>                          
#> 1 Not distance-education only  = 0
#> 2 Distance-education only  = 1
```

<!-- In binding multiple years of data with `dyplr::bind_rows()`, however, labels will get dropped. To throw back the labels, `scorecard` also provides to simple functions that attach variable and value labels to the tibbles: -->

<!-- ```{r} -->

<!-- bind_rows( -->

<!--   scorecard::mf2014_15 %>% select(mf_year, instnm, control, adm_rate), -->

<!--   scorecard::mf2013_14 %>% select(mf_year, instnm, control, adm_rate) -->

<!-- ) %>%  -->

<!--   scorecard::attach_var_label() %>%  -->

<!--   scorecard::attach_val_label() -> df -->

<!-- df -->

<!-- labelled::var_label(df) -->

<!-- labelled::val_labels(df) -->

<!-- ``` -->

## Exploring codebook and plot distributions of in-state tuition

``` r
vars = c("mf_year", "iclevel", "control", "tuitionfee_in")

scorecard::codebook %>% 
  select(var_name, var_label, value, val_label) %>% 
  filter(var_name %in% vars) %>% 
  knitr::kable()
```

| var\_name      | var\_label                | value | val\_label         |
| :------------- | :------------------------ | ----: | :----------------- |
| control        | Control of institution    |     1 | Public             |
| control        | Control of institution    |     2 | Private nonprofit  |
| control        | Control of institution    |     3 | Private for-profit |
| tuitionfee\_in | In-state tuition and fees |    NA | NA                 |
| iclevel        | Level of institution      |     1 | 4-year             |
| iclevel        | Level of institution      |     2 | 2-year             |
| iclevel        | Level of institution      |     3 | Less-than-2-year   |

``` r

dplyr_seq = . %>% 
  select(one_of(vars)) %>%
  haven::as_factor() %>% 
  filter(iclevel %in% c("4-year", "2-year")) %>% 
  mutate(year = mf_year %>% parse_number() %>% as.factor()) %>% 
  group_by(iclevel, control) %>% 
  mutate_at(c("tuitionfee_in"),
            ~statar::winsorise(., probs = c(0.02, 0.98), verbose = FALSE)) %>% 
  ungroup()

## Test the functional sequence
scorecard::mf2014_15 %>% dplyr_seq()
#> # A tibble: 5,530 x 5
#>    mf_year iclevel control           tuitionfee_in year 
#>    <chr>   <fct>   <fct>                     <dbl> <fct>
#>  1 2014-15 4-year  Public                     9096 2014 
#>  2 2014-15 4-year  Public                     7510 2014 
#>  3 2014-15 4-year  Private nonprofit          6900 2014 
#>  4 2014-15 4-year  Public                     9158 2014 
#>  5 2014-15 4-year  Public                     8720 2014 
#>  6 2014-15 4-year  Public                     9826 2014 
#>  7 2014-15 2-year  Public                     3491 2014 
#>  8 2014-15 4-year  Public                       NA 2014 
#>  9 2014-15 4-year  Public                     9080 2014 
#> 10 2014-15 4-year  Public                    10200 2014 
#> # … with 5,520 more rows

bind_rows(
  scorecard::mf2017_18 %>% dplyr_seq(),
  scorecard::mf2016_17 %>% dplyr_seq(),
  scorecard::mf2015_16 %>% dplyr_seq(),
  scorecard::mf2014_15 %>% dplyr_seq(),
  scorecard::mf2013_14 %>% dplyr_seq(),
  scorecard::mf2012_13 %>% dplyr_seq(),
  scorecard::mf2011_12 %>% dplyr_seq(),
  scorecard::mf2010_11 %>% dplyr_seq(),
  scorecard::mf2009_10 %>% dplyr_seq(),
  scorecard::mf2008_09 %>% dplyr_seq(),
  scorecard::mf2007_08 %>% dplyr_seq(),
  scorecard::mf2006_07 %>% dplyr_seq(),
  scorecard::mf2005_06 %>% dplyr_seq(),
  scorecard::mf2004_05 %>% dplyr_seq(),
  scorecard::mf2003_04 %>% dplyr_seq(),
  scorecard::mf2002_03 %>% dplyr_seq(),
  scorecard::mf2001_02 %>% dplyr_seq(),
  scorecard::mf2000_01 %>% dplyr_seq()
) -> df

df %>% 
  ggplot(aes(x = tuitionfee_in, y = year, fill = iclevel)) +
  ggjoy::geom_joy(scale = 2, alpha = .8, colour = "white") +
  ggjoy::theme_joy() +
  facet_grid(iclevel~control, scales = "free") +
  labs(x = NULL, y = NULL,
       title = "In-State Tuition and Fees, 2000-2017") +
  scale_x_continuous(labels = scales::dollar) +
  scale_y_discrete(breaks = seq(2017, 2000, -3), 
                   expand = c(0.01, 0)) +
  theme(axis.text = element_text(size = 8),
        legend.position = "none")
```

![](README-files/in_tuition-1.png)<!-- -->

## Compareing in-state and out-of-state tuition and fees

![](README-files/in_or_out_tuition-1.png)<!-- -->

## Distribution of Average Age of Entry

![](README-files/age-of-entry-1.png)<!-- -->![](README-files/age-of-entry-2.png)<!-- -->


NEWS.md

# `scorecard` 0.4.0

* Added 2016-17 data
* Added 2017-18 data

# `scorecard` 0.3.0

* Added 2015-16 data

# `scorecard` 0.2.1

* Drop empty (all NAs) columns for all datasets

# `scorecard` 0.2.0

* Added a `NEWS.md` file to track changes to the package
* Updated to 2014-15 data


R/data.R
#' College Scorecard Data, 1996-97
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf1996_97"

#' College Scorecard Data, 1997-98
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf1997_98"

#' College Scorecard Data, 1998-99
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf1998_99"

#' College Scorecard Data, 1999-00
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf1999_00"

#' College Scorecard Data, 2000-01
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2000_01"

#' College Scorecard Data, 2001-02
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2001_02"

#' College Scorecard Data, 2002-03
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2002_03"

#' College Scorecard Data, 2003-04
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2003_04"

#' College Scorecard Data, 2004-05
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2004_05"

#' College Scorecard Data, 2005-06
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2005_06"

#' College Scorecard Data, 2006-07
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2006_07"

#' College Scorecard Data, 2007-08
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2007_08"

#' College Scorecard Data, 2008-09
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2008_09"

#' College Scorecard Data, 2009-10
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2009_10"

#' College Scorecard Data, 2010-11
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2010_11"

#' College Scorecard Data, 2011-12
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2011_12"

#' College Scorecard Data, 2012-13
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2012_13"

#' College Scorecard Data, 2013-14
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2013_14"

#' College Scorecard Data, 2014-15
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2014_15"

#' College Scorecard Data, 2015-16
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2015_16"

#' College Scorecard Data, 2016-17
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2016_17"

#' College Scorecard Data, 2017-18
#'
#' Contains all the variables from College Scorecard Merged Files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"mf2017_18"

#' College Scorecard Data codebook
#'
#' Contains variable names, labels, label values for the College Scorecard Data.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"codebook"

#' Cohort Map
#'
#' Provides descriptions of which years of data are described for each variable
#' in the merged files.
#'
#' @md
#' @source [https://collegescorecard.ed.gov/data]()
"cohort_map"


data-raw/prepare-data.R
# Load libraries ----------------------------------------------------------

rm(list = ls())
library(readr)
library(dplyr)
library(devtools)

# Download the raw data ---------------------------------------------------

download.file("https://s3.amazonaws.com/ed-college-choice-public/CollegeScorecard_Raw_Data.zip",
              "data-raw/raw.zip")

# Save the datasets -------------------------------------------------------

merged1996 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED1996_PP.csv"))
use_data(merged1996)

merged1997 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED1997_PP.csv"))
use_data(merged1997)

merged1998 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED1998_PP.csv"))
use_data(merged1998)

merged1999 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED1999_PP.csv"))
use_data(merged1999)

merged2000 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2000_PP.csv"))
use_data(merged2000)

merged2001 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2001_PP.csv"))
use_data(merged2001)

merged2002 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2002_PP.csv"))
use_data(merged2002)

merged2003 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2003_PP.csv"))
use_data(merged2003)

merged2004 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2004_PP.csv"))
use_data(merged2004)

merged2005 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2005_PP.csv"))
use_data(merged2005)

merged2006 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2006_PP.csv"))
use_data(merged2006)

merged2007 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2007_PP.csv"))
use_data(merged2007)

merged2008 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2008_PP.csv"))
use_data(merged2008)

merged2009 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2009_PP.csv"))
use_data(merged2009)

merged2010 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2010_PP.csv"))
use_data(merged2010)

merged2011 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2011_PP.csv"))
use_data(merged2011)

merged2012 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2012_PP.csv"))
use_data(merged2012)

merged2013 = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/MERGED2013_PP.csv"))
use_data(merged2013)

codebook = read_csv(unz("data-raw/raw.zip", filename = "CollegeScorecard_Raw_Data/CollegeScorecardDataDictionary-09-12-2015.csv"))
use_data(codebook)

# Delete the raw zip file -------------------------------------------------

# unlink("data-raw/raw.zip")



data-raw/prep.R

library(tidyverse)
library(devtools)
library(stringr)
library(readxl)
library(labelled)
library(containerit)

# Save the codebook and cohort_map files ----------------------------------
# Include data from the most recent update (December 12, 2019)

url = "https://collegescorecard.ed.gov/assets/CollegeScorecardDataDictionary.xlsx"
fil = "data-raw/CollegeScorecardDataDictionary_122019.xlsx"
if (!file.exists(fil)) download.file(url, fil)

read_excel(fil, sheet = 4) %>%
  rename_all(tolower) %>%
  rename(var_name = `variable name`,
         var_label = `name of data element`,
         dev_category = `dev-category`,
         val_label = label,
         dev_friendly_name = `developer-friendly name`,
         api_data_type = `api data type`) %>%
  select(var_name, var_label, dev_category,
         value, val_label, source, everything()) %>%
  mutate(var_name = tolower(var_name),
         var_label = gsub("[\r\n]", "", var_label)) %>%
  fill(var_name, var_label, source, dev_friendly_name, api_data_type) %>%
  print() -> codebook

readxl::read_excel(fil, sheet = 5) %>%
  rename_all(tolower) %>%
  rename(var_name = `variable name`) %>%
  mutate(var_name = tolower(var_name)) %>%
  gather(datafile, note, -var_name) %>%
  print() -> cohort_map

use_data(codebook, cohort_map, overwrite = TRUE)

# Prepare variable label and value labels ---------------------------------

codebook %>%
  select(var_name, var_label) %>%
  distinct(var_name, .keep_all = TRUE) %>%
  deframe() %>%
  as.list() %>%
  print() -> var_label_lst

codebook %>%
  select(var_name, val_label, value) %>%
  filter(!is.na(value)) %>%
  group_by(var_name) %>%
  nest() %>%
  mutate(data = map(data, deframe)) %>%
  deframe() %>%
  print() -> val_label_lst

use_data(var_label_lst, val_label_lst, overwrite = TRUE, internal = TRUE)

# Download the raw data ---------------------------------------------------
# Include data from the most recent update (December 12, 2019)

url = "https://ed-public-download.app.cloud.gov/downloads/CollegeScorecard_Raw_Data.zip"
zip = "data-raw/CollegeScorecard_Raw_Data_122019.zip"
if (!file.exists(zip)) download.file(url, zip)

# List files in the zip file ----------------------------------------------

unzip(zip, list = TRUE) %>%
  as_tibble() %>%
  arrange(Name) %>%
  print(n = 28)

# # A tibble: 28 x 3
#   Name                                                           Length Date
#   <chr>                                                           <dbl> <dttm>
# 1 CollegeScorecard_Raw_Data/                                          0 2019-12-03 16:12:00
# 2 CollegeScorecard_Raw_Data/.DS_Store                              6148 2019-11-18 21:06:00
# 3 CollegeScorecard_Raw_Data/Crosswalks.zip                     35375490 2019-05-06 07:06:00
# 4 CollegeScorecard_Raw_Data/data.yaml                            486192 2019-10-13 13:13:00
# 5 CollegeScorecard_Raw_Data/FieldOfStudyData1415_1516_PP.csv   43757585 2019-11-15 15:14:00
# 6 CollegeScorecard_Raw_Data/FieldOfStudyData1516_1617_PP.csv   51099892 2019-11-15 15:14:00
# 7 CollegeScorecard_Raw_Data/MERGED1996_97_PP.csv               63348240 2019-10-10 09:18:00
# 8 CollegeScorecard_Raw_Data/MERGED1997_98_PP.csv               66766520 2019-10-10 09:19:00
# 9 CollegeScorecard_Raw_Data/MERGED1998_99_PP.csv               77519889 2019-10-10 09:19:00
# 10 CollegeScorecard_Raw_Data/MERGED1999_00_PP.csv              91808656 2019-10-10 09:20:00
# 11 CollegeScorecard_Raw_Data/MERGED2000_01_PP.csv             104425976 2019-10-10 09:21:00
# 12 CollegeScorecard_Raw_Data/MERGED2001_02_PP.csv             107365717 2019-10-10 09:21:00
# 13 CollegeScorecard_Raw_Data/MERGED2002_03_PP.csv             118529545 2019-10-10 09:22:00
# 14 CollegeScorecard_Raw_Data/MERGED2003_04_PP.csv             120662741 2019-10-10 09:23:00
# 15 CollegeScorecard_Raw_Data/MERGED2004_05_PP.csv             133477657 2019-10-10 09:24:00
# 16 CollegeScorecard_Raw_Data/MERGED2005_06_PP.csv             137842770 2019-10-10 09:24:00
# 17 CollegeScorecard_Raw_Data/MERGED2006_07_PP.csv             136534126 2019-10-10 09:25:00
# 18 CollegeScorecard_Raw_Data/MERGED2007_08_PP.csv             139346257 2019-10-10 09:26:00
# 19 CollegeScorecard_Raw_Data/MERGED2008_09_PP.csv             139318440 2019-10-10 09:27:00
# 20 CollegeScorecard_Raw_Data/MERGED2009_10_PP.csv             145004052 2019-10-10 09:28:00
# 21 CollegeScorecard_Raw_Data/MERGED2010_11_PP.csv             148149393 2019-10-10 09:28:00
# 22 CollegeScorecard_Raw_Data/MERGED2011_12_PP.csv             155062442 2019-10-10 09:29:00
# 23 CollegeScorecard_Raw_Data/MERGED2012_13_PP.csv             157050875 2019-10-10 09:30:00
# 24 CollegeScorecard_Raw_Data/MERGED2013_14_PP.csv             157811300 2019-10-10 09:31:00
# 25 CollegeScorecard_Raw_Data/MERGED2014_15_PP.csv             156335981 2019-10-10 09:32:00
# 26 CollegeScorecard_Raw_Data/MERGED2015_16_PP.csv             152061339 2019-10-10 09:33:00
# 27 CollegeScorecard_Raw_Data/MERGED2016_17_PP.csv             153213137 2019-09-18 18:56:00
# 28 CollegeScorecard_Raw_Data/MERGED2017_18_PP.csv              68092022 2019-12-02 06:59:00

# Save main datasets -------------------------------------------------------

## List of csv files
unzip(zip, list = TRUE) %>%
  as_tibble() %>%
  arrange(Name) %>%
  filter(grepl("MERGED", Name)) %>%
  pull(Name) %>%
  print() -> csv_lst

## Dataset names
csv_lst %>%
  str_extract("([0-9]{1,4}_[0-9]{1,2})") %>%
  paste0("mf", .) %>%
  print() -> data_names

## Merged file years
csv_lst %>%
  str_extract("([0-9]{1,4}_[0-9]{1,2})") %>%
  str_replace("_", "-") %>%
  print() -> mf_year

## Load csv files, assign var labels, val labels, and dataset names
csv_lst %>%
  map(~unz(zip, .)) %>%
  map(~read_csv(., guess_max = 10000, na = c("", "NA", "NULL"))) %>%
  map(~rename_all(., tolower)) %>%
  ### Add mf_year as the first column
  map2(.y = mf_year, ~mutate(., mf_year = .y)) %>%
  map(~select(., mf_year, everything())) %>%
  ### Identifiers should always be string
  map(~mutate_at(., c("unitid", "opeid", "opeid6"), as.character)) %>%
  ### Variables that has value labels should be numeric
  map(~mutate_at(., names(val_label_lst), as.numeric)) %>%
  map(~`var_label<-`(., var_label_lst)) %>%
  map(~`val_labels<-`(., val_label_lst)) %>%
  ### Drop columns that all values are NA
  map(~discard(., ~all(is.na(.x)))) %>%
  ### Assign dataset names
  map2(.y = data_names, ~assign(.y, .x, envir = .GlobalEnv))

use_data(mf1996_97, overwrite = TRUE)
use_data(mf1997_98, overwrite = TRUE)
use_data(mf1998_99, overwrite = TRUE)
use_data(mf1999_00, overwrite = TRUE)
use_data(mf2000_01, overwrite = TRUE)
use_data(mf2001_02, overwrite = TRUE)
use_data(mf2002_03, overwrite = TRUE)
use_data(mf2003_04, overwrite = TRUE)
use_data(mf2004_05, overwrite = TRUE)
use_data(mf2005_06, overwrite = TRUE)
use_data(mf2006_07, overwrite = TRUE)
use_data(mf2007_08, overwrite = TRUE)
use_data(mf2008_09, overwrite = TRUE)
use_data(mf2009_10, overwrite = TRUE)
use_data(mf2010_11, overwrite = TRUE)
use_data(mf2011_12, overwrite = TRUE)
use_data(mf2012_13, overwrite = TRUE)
use_data(mf2013_14, overwrite = TRUE)
use_data(mf2014_15, overwrite = TRUE)
use_data(mf2015_16, overwrite = TRUE)
use_data(mf2016_17, overwrite = TRUE)
use_data(mf2017_18, overwrite = TRUE)

# Delete the raw zip file -------------------------------------------------
# unlink(zip)

# Container it ------------------------------------------------------------

container = dockerfile(from = sessionInfo(),
                       maintainer = "jjchern",
                       r_version = "3.6.1")
write(container, "data-raw/prep_container.dockerfile")
